spec/src/main/asciidoc/cdi.asciidoc

```diff
@@ -19,4 +19,158 @@
+
+  @Inject
+  @NamedInstance("myExecutor")
+  ManagedExecutor sharedExecutor;
+
+}
+----
+
+Since the configuration of `ManagedExecutor` is optional, a shared `ManagedExecutor` with default configuration is also a possibility.
+
+[source, java]
+----
+@ApplicationScoped
+public class FooBean {
+
+  @Inject
+  @NamedInstance("sharedDefaultME")
+  ManagedExecutor executor;
+
+}
+
+@ApplicationScoped
+public class BarBean {
+
+  @Inject
+  @NamedInstance("sharedDefaultME")
+  ManagedExecutor sharedExecutor;
+
+}
+----
+
+If multiple configurations of `ManagedExecutor` are assigned the same name, the container detects it and raises a `javax.enterprise.inject.spi.DefinitionException`.
+
+=== Injecting `ThreadContext`
+
+Application can inject instances of `ThreadContext` using CDI with the option to share same `ThreadContext` between multiple injection points.
+The simplest use case will create a new `ThreadContext` instance for every such injection point and the configuration will be equal to default values of `@ThreadContextConfig` annotation.
+
+[source, java]
+----
+@Inject
+ThreadContext context;
+----
+
+==== Configuring `ThreadContext`
+
+Oftentimes there is need to fine-grain the `ThreadContext` instance and modify what contexts are to be propagated.
+This is where `@ThreadContextConfig` annotation comes into play allowing to configure all the aspects of an injected `ThreadContext` instance.
+Note that using just this annotation, new instance of `ThreadContext` is created for every injection point even should those injection points have equal configuration.
+
+[source, java]
+----
+@Inject
+@ThreadContextConfig(propagated = ThreadContext.SECURITY)
+ThreadContext configuredContext;
+----
+
+==== Sharing `ThreadContext` Instances
+
+Last but not least, sometimes it is desirable to share the same instance of `ThreadContext` among several injection points.
+For such use case, the specification provides a CDI qualifier `@NamedInstance` which takes one parameter - a unique name assigned to the given `ThreadContext` configuration.
+Having one injection point that defines both, the name using `@NamedInstance` and the configuration using `@ThreadContextConfig`, application can then inject the same instance repeatedly using the qualifier.
+Assuming two beans wish to share a `ThreadContext` instance, the code looks as follows:
+
+[source, java]
+----
+@ApplicationScoped
+public class MyFirstBean {
+
+  @Inject
+  @NamedInstance("myThreadContext")
+  @ThreadContextConfig(propagated = ThreadContext.SECURITY)
+  ThreadContext context;
+
+}
+
+@ApplicationScoped
+public class MySecondBean {
+
+  @Inject
+  @NamedInstance("myThreadContext")
+  ManagedExecutor sharedContext;
+
+}
+----
+
+Since the configuration of `ThreadContext` is optional, a shared `ThreadContext` with default configuration is also a possibility.
+
+[source, java]
+----
+@ApplicationScoped
+public class FooBean {
+
+  @Inject
+  @NamedInstance("sharedDefaultTC")
```

Comment:
same comment here as with ManagedExecutors.  Also, if we are going to show examples of invalid code in our spec doc, we should make it more obvious that the code example is bad, or point out via in-line comments what is wrong with the code example.  
Reading this, I'm not certain if the following text about a DefinitionException is describing this code example or not.

Reply:
> Reading this, I'm not certain if the following text about a DefinitionException is describing this code example or not.

All proposed text describes how to achieve a certain behaviour and then shows it with bits of code and this is no exception. Besides this code won't throw the exception because similar configuration is not given a different name.

Reply:
Note: This falls under the same category as the similar section for ManagedExecutor that will be discussed under the main issue